### PR TITLE
[v2] disperser client payments api

### PR DESCRIPTION
**Pull request by hopeyen**
## Why are these changes needed?

properly construct payment header in the dispersal request

For easy initialization of DisperserClient, pass in an empty accountant `&accountant{}` and then call `client.PopulateAccountant(ctx)` to populate the accountant by getting payment information both on-chain and off-chain for the local signer.

For more customized initialization, the user can explicitly construct their accountant and pass in as the argument

## Checks

- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- [ ] I've checked the new test coverage and the coverage percentage didn't drop.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by hopeyen** [comment]
Waiting for #902 to merge before rebasing

File: api/clients/disperser_client_v2.go
Comment (ian-shim) [inline api/clients/disperser_client_v2.go:63]: It's quite difficult to construct a new Accountant. Does it make sense to provide a helper method that creates an accountant from the `GetPaymentState` result? 

Let's also add a way to call `GetPaymentState` in this client

**Review by ian-shim — APPROVED** [review]

File: api/clients/disperser_client_v2.go
Comment (hopeyen) [reply api/clients/disperser_client_v2.go]: added a helper function `PopulateAccountant` that should run after creating `NewDisperserClientV2` by constructing accountant from `GetPaymentState`

What do you think about allowing the user of DisperserClientV2 to provide their own accountant, or simply passing in `&accountant{}` to `NewDisperserClientV2` and then explicitly calling `PopulateAccountant` to fill in the accountant state? ([5004cf9](https://github.com/Layr-Labs/eigenda/pull/928/commits/5004cf9ad401f641afce86088fd2633d1b947f87))

Comment (hopeyen) [reply api/clients/disperser_client_v2.go]: The reason is `GetPaymentState` is a function of `disperserClientV2`. If we write a standalone `PopulateAccountant`, then we would need it to be able to `initOnceGrpcConnection`, pass in a signer for creating the request, and a `disperser_rpc.DisperserClient` that makes the actual request. I figured it is cleaner to just use disperserClientV2

File: api/clients/disperser_client_v2.go
Comment (ian-shim) [inline api/clients/disperser_client_v2.go:123]: check if `c.accountant` is nil? 
we can instruct how to initialize accountant in the error message

File: api/clients/disperser_client_v2.go
Comment (ian-shim) [inline api/clients/disperser_client_v2.go:87]: This function isn't required if a valid accountant is passed in the constructor, right?
If you want to require this method, we can remove accountant as an argument from the constructor.

File: api/clients/accountant.go
Comment (ian-shim) [inline api/clients/accountant.go:163]: these will panic if any of the attributes are nil?

Comment (hopeyen) [reply api/clients/accountant.go]: yes.... added a check for all of them!
It would've been great if protobuf let me set fields to be required, but proto3 removed "required" keyword:(

File: api/clients/disperser_client_v2.go
Comment (hopeyen) [reply api/clients/disperser_client_v2.go]: Oh yeah, will update the comment

File: api/clients/accountant.go
Comment (ian-shim) [reply api/clients/accountant.go]: If you use getters (i.e. `GetXXX`), you could chain them!

**Review by ian-shim — APPROVED** [review]
lgtm, lets address the lint issue

File: api/clients/accountant.go
Comment (samlaf) [inline api/clients/accountant.go:162]: This feels too coupled to me. A Public setter that is not part of the interface and it set by some response from another client. As discussed in other threads, could we not combine accountant and disperser_client more smoothly? Does accountant really need to be an independent struct?

File: api/clients/accountant.go
Comment (samlaf) [inline api/clients/accountant.go:20]: Why is this interface defined if it's not being used in the disperser_client?

File: api/clients/disperser_client_v2.go
Comment (samlaf) [inline api/clients/disperser_client_v2.go:127]: how about we make the call to PopulateAccountant as part of the initOnce function? This way its hidden from the user.
Is there ever a need to recall PopulateAccountant? For eg if disperser went down, or whatever other weird case? To reset the accountant basically? If so, should it be called `ResetAccountantFromServer` or something like that?

File: api/clients/accountant.go
Comment (hopeyen) [reply api/clients/accountant.go:162]: We wanted to keep accounting flow separate from the client, but arguably we can combine them into a bigger struct. What do you think @ian-shim ?

File: api/clients/disperser_client_v2.go
Comment (hopeyen) [reply api/clients/disperser_client_v2.go:127]: In this case, maybe we also move signer check to the initOnce?

There's some cases a user might recall populate accountant, I don't think it hurts to be able to reset. Can update the function name for sure

File: api/clients/accountant.go
Comment (hopeyen) [reply api/clients/accountant.go:20]: `AccountBlob` is used in `DisperseBlob`, but I guess we wouldn't need the interface to access this function... I thought this is good practice, but happy to remove if that's not the case

Comment (hopeyen) [reply api/clients/accountant.go:20]: Makes sense. Can update the names from `Accountant-> IAccountant` for interface and `accountant-> Accountant`

Comment (samlaf) [reply api/clients/accountant.go:20]: Think `Accounter` would be more golangy? But in either case, Unless this interface is already used in many places, best practice is to just delete it. Interfaces should typically be defined at the call site, not where the struct is defined. That's because different call sites need different subsets of methods of a struct typically, so best to define the minimal interface needed (makes it easier to write a fake for that minimal interface).

Comment (hopeyen) [reply api/clients/accountant.go:20]: I removed the interface definition and renamed the struct to Accounter

File: api/clients/disperser_client_v2.go
Comment (hopeyen) [reply api/clients/disperser_client_v2.go:127]: I think I'll leave this as a future refactoring

File: api/clients/accountant.go
Comment (samlaf) [reply api/clients/accountant.go:20]: Wah sorry! The struct should still be called Accountant. I was saying IF we were to have an interface here (which I don't think we should), then it could be called `Accounter` instead of `IAccountant`. It's typical of golang to use verbs for interface, like Writer and Reader.

Comment (samlaf) [reply api/clients/accountant.go]: LGTM after [6643ca9](https://github.com/Layr-Labs/eigenda/pull/928/commits/6643ca9280c1c4d8cd7c59cde05a939e9813a5eb); Ill resolve this

**Review by samlaf — APPROVED** [review]
I'll give a stamp on this in case its blocking progress on other work, but I would have liked to get @ian-shim's take on https://github.com/Layr-Labs/eigenda/pull/928#discussion_r1872241197. Would really like to find a better interface/decoupling between these. But we can do so later (hopefully...)

File: api/clients/accountant.go
Comment (samlaf) [reply api/clients/accountant.go]: Makes sense! I would also not try to future-proof this design too much (to a case where there are multiple dispersers and you need some super complex accountant that manages multiple dispersers). We should try to make the simple case simple, true to golang's philosophy, and people who need complex integrations can take on the complexity hit around our simple API.

Another possibility is to have 2 constructors: one where we inject the accountant, and the other that initializes the accountant by calling the disperser?

Comment (hopeyen) [reply api/clients/accountant.go]: Hmm all one needs to do is to call the current constructor and then call `PopulateAccountant`, so adding a separate constructor doesn't seem that beneficial to me. I won't mind adding it, probably in the follow-up [PR](https://github.com/Layr-Labs/eigenda/pull/965) that focus on integrations adaptions (maybe something like `NewDisperserClientV2AutoPay`?

**Pull request merged**
